compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionPreprocessing.cpp

```diff
@@ -142,6 +142,104 @@ struct FoldSuccessiveTensorInsertSliceOps
   }
 };
 
+//===----------------------------------------------------------------------===//
+// ExtractAtSmallBitwidth
+//===----------------------------------------------------------------------===//
+
+///
+/// ```
+/// %0 = linalg.generic i8 -> f32
+/// %1 = tensor.extract_slice
+/// %2 = linalg.generic f32 -> i8
+/// ```
+///
+/// to
+///
+/// ```
+/// %0 = linalg.generic i8 -> f32
+/// %new1 = linalg.generic f32 -> i8
+/// %1 = tensor.extract_slice i8
+/// %new2 = linalg.generic i8 -> f32
+/// %2 = linalg.generic f32 -> i8
+/// ```
+struct ExtractAtSmallBitwidth
+    : public OpRewritePattern<tensor::ExtractSliceOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensor::ExtractSliceOp sliceOp,
+                                PatternRewriter &rewriter) const override {
+
+    auto sourceGenericOp =
+        sliceOp.getSource().getDefiningOp<linalg::GenericOp>();
+    if (!sourceGenericOp || !isDequantizationLikeOp(sourceGenericOp)) {
+      return rewriter.notifyMatchFailure(
+          sliceOp, "expected source to be dequantize-like generic op");
+    }
+
+    auto sourceResult = sourceGenericOp.getResult(0);
+    auto sourceNarrowType = dyn_cast<RankedTensorType>(sourceResult.getType());
+    auto sourceWideType = cast<RankedTensorType>(
+        sourceGenericOp.getDpsInputs().front().getType());
+    auto extractedNarrowType = RankedTensorType::get(
+        sliceOp.getResultType().getShape(), sourceNarrowType.getElementType());
+
+    // if (!sourceWideType.getElementType().isF32() ||
+    //     !sourceNarrowType.getElementType().isInteger() ||
+    //     sourceNarrowType.getElementType().getIntOrFloatBitWidth() >=
+    //         sourceWideType.getElementType().getIntOrFloatBitWidth()) {
+    //   return rewriter.notifyMatchFailure(
+    //       sliceOp, "expected int to float conversion with widening
+    //       bitwidth");
+    // }
+
+    // Create a `linalg.generic` to narrow the element bitwidth back to the
+    // original size
+    Value emptyOp = rewriter.create<tensor::EmptyOp>(
+        sliceOp.getLoc(), sourceWideType.getShape(),
+        sourceWideType.getElementType());
+
+    auto newGeneric = rewriter.create<linalg::GenericOp>(
+        sliceOp.getLoc(), sourceWideType, sourceGenericOp.getResults(), emptyOp,
+        sourceGenericOp.getIndexingMapsArray(),
+        sourceGenericOp.getIteratorTypesArray(),
+        [&](OpBuilder &nestedBuilder, Location loc, ValueRange args) {
+          // Custom region for f32 -> i8 conversion
+          auto castOp = nestedBuilder.create<arith::FPToSIOp>(
```

Comment:
How does modification by the quantization scale come into play here? If you just cast the fp to si, I can't imagine you will retain any useful information. 

Is it possible to copy the payload of a generic op, but modify the other information?

Reply:
Yeah your right, this is definitely a problem. I think that solution might work, I'll take a look